### DISCUSSION TITLE
Several drawing questions

### DISCUSSION THREAD

**Discussion by Antix-Development** [mono/SkiaSharp#2453] [Q&A] [answered]
I'm writing an Avalonia application where I am using a custom control that contains an SKCanvas.

1. Is it possible to draw directly from one SKCanvas to another SKCanvas? I cannot seem to be able to find any information that says it is possible.

2. I'd also like to save an SKCanvas as a PNG file but I also cannot find anything that hints that it is possible.

I'm already thinking I am going to have to ditch the SKCanvas control and find (or fabricate) something different, maybe using SKBitmaps. I'd just rather not have to :)

**Comment by taublast** [comment, marked as answer]
1 you can write from one canvas' surface to another canvas,:

`srcSurface.Draw(destSurface.Canvas, destinationLeft, destinationTop, paint); //if you set paint color to white with alpha, you will draw with opacity of that alpha, like a semi-transparent layer`

watch for size mismatch and commit changes `srcSurface.Canvas.Flush()` before copying if needed.

2 the logic when saving surface to file could be:

```
public Stream ToStream(SKSurface surface, SKEncodedImageFormat format = SKEncodedImageFormat.Jpeg, int quality = 90)
		{
			try
			{

				using (var image = surface.Snapshot())
				{
					var bitmap = SKBitmap.FromImage(image);
					var rotated = Reorient(bitmap);

					var data = rotated.Encode(format, quality);
					return data.AsStream();
				}

				SKBitmap Reorient(SKBitmap bitmap)
				{
					SKBitmap rotated;
					switch (SensorRotation)
					{
						case 180:
							using (var surface = new SKCanvas(bitmap))
							{
								surface.RotateDegrees(180, bitmap.Width / 2.0f, bitmap.Height / 2.0f);
								surface.DrawBitmap(bitmap.Copy(), 0, 0);
							}
							return bitmap;
						case 270:
							rotated = new SKBitmap(bitmap.Height, bitmap.Width);
							using (var surface = new SKCanvas(rotated))
							{
								surface.Translate(rotated.Width, 0);
								surface.RotateDegrees(90);
								surface.DrawBitmap(bitmap, 0, 0);
							}
							return rotated;
						case 90:
							rotated = new SKBitmap(bitmap.Height, bitmap.Width);
							using (var surface = new SKCanvas(rotated))
							{
								surface.Translate(0, rotated.Height);
								surface.RotateDegrees(270);
								surface.DrawBitmap(bitmap, 0, 0);
							}
							return rotated;
						default:
							return bitmap;
					}
				}
			}
			catch (Exception e)
			{
				Console.WriteLine(e);
				return null;
			}
		}
```

output stream to file at will.

**Reply by Antix-Development** [reply]
Thanks, that's extremely helpful :)
